ADR 038: State Listening (#8012)
cosmos/cosmos-sdk: docs/architecture/adr-038-state-listening.md

# ADR 038: KVStore state listening

## Changelog

- 11/23/2020: Initial draft

## Status

Proposed

## Abstract

This ADR defines a set of changes to enable listening to state changes of individual KVStores and exposing these data to consumers.

## Context

Currently, KVStore data can be remotely accessed through [Queries](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/messages-and-queries.md#queries) which proceed through Tendermint and the ABCI.
In addition to these request/response queries, it would be beneficial to have a means of listening to state changes as they occur in real time.

## Decision

We will modify the `MultiStore` interface and its concrete (`rootmulti` and `cachemulti`) implementations and introduce a new `listenkv.Store` to allow listening to state changes in underlying KVStores.
We will also introduce two approaches for exposing the data to external consumers: writing to files and writing to a gRPC stream.

> ### Listening interface

Reviewer: ```suggestion
### Listening interface

```

In a new file- `store/types/listening.go`- we will create a `WriteListener` interface for streaming out state changes from a KVStore.

```go

> // WriteListener interface for writing data out from a listenkv.Store

Reviewer: From reading and understanding this, `WriteListener` is more of a streaming proxy from a producer (module KVStore) to a consumer (essentially any `io.Writer`), there inherently does not need to be any writing on the consumer-side, although obviously there will be. I would rephrase this godoc.

Author: Absolutely, "streaming data out" would be more accurate.

type WriteListener interface {
  // if value is nil then it was deleted
  OnWrite(key []byte, value []byte)
}
```

> ### Listener type

Reviewer: ```suggestion
### Listener type

```

We will create a concrete implementation of the `WriteListener` interface in `store/types/listening.go` that gob

> encodes and writes the key value pair into an underlying `io.Writer`.

Reviewer: Why gob? Why not just length-prefix key/values with their raw bytes since that's what we're proxying anyway? I ask because the true consumer (typically an `io.Writer`) won't necessary be Go-based.

Better to provide a concrete type that doesn't depend on language-specific encoding.

Author: Gob was chosen somewhat arbitrarily for its efficiency and simplicity, will adjust this to be language agnostic!

Author: We are using protobuf type now

```go
// GobWriteListener is used to configure listening to a KVStore using a gob encoding wrapper around an io.Writer
type GobWriteListener struct {
	writer              *gob.Encoder
}

// NewGobWriteListener wraps a WriteListenerWriter around an io.Writer
func NewGobWriteListener(w io.Writer) *GobWriteListener {
	return &GobWriteListener{
		writer: gob.NewEncoder(w),
	}
}

// KVPair for gob encoding
type KVPair struct {
	Key []byte
	Value []byte
}

// OnWrite satisfies WriteListener interface by writing out key-value gobs to the underlying io.Writer
func (l *Listener) OnWrite(key []byte, value []byte) {
	l.writer.Encode(KVPair{Key: key, Value: value})
}
```

> ### ListenKVStore
> We will create a new `Store` type `listenkv.Store` that the `MultiStore` wraps around a `KVStore` to enable state listening.

Reviewer: ```suggestion
### ListenKVStore

We will create a new `Store` type `listenkv.Store` that the `MultiStore` wraps around a `KVStore` to enable state listening.
```

// Operations are traced on each core KVStore call and written to any of the
// underlying listeners with the proper key and operation permissions
type Store struct {
    parent    types.KVStore
    listeners []types.WriteListener
}

// NewStore returns a reference to a new traceKVStore given a parent
// KVStore implementation and a buffered writer.
func NewStore(parent types.KVStore, listeners []types.WriteListener) *Store {
	return &Store{parent: parent, listeners: listeners}
}

// Set implements the KVStore interface. It traces a write operation and
// delegates the Set call to the parent KVStore.
func (tkv *Store) Set(key []byte, value []byte) {
	types.AssertValidKey(key)
	onWrite(tkv.listeners, key, value)
	tkv.parent.Set(key, value)
}

// Delete implements the KVStore interface. It traces a write operation and
// delegates the Delete call to the parent KVStore.
func (tkv *Store) Delete(key []byte) {
	onWrite(tkv.listeners, key, nil)
	tkv.parent.Delete(key)
}

// onWrite writes a KVStore operation to all of the WriteListeners
func onWrite(listeners []types.WriteListener, key, value []byte) {
	for _, l := range listeners {
		l.OnWrite(key, value)
	}
}
```

> ### MultiStore interface updates

Reviewer: ```suggestion
### MultiStore interface updates

```

We will update the `MultiStore` interface to allow us to wrap a set of listeners around a specific `KVStore`.
Additionally, we will update the `CacheWrap` and `CacheWrapper` interfaces to enable listening in the caching layer, and add a `MultiStore` method
to turn on or off this cache listening.

```go
type MultiStore interface {
	...

	// ListeningEnabled returns if listening is enabled for the KVStore belonging the provided StoreKey
	ListeningEnabled(key StoreKey) bool

	// SetListeners sets the WriteListeners for the KVStore belonging to the provided StoreKey
	SetListeners(key StoreKey, listeners []WriteListener)

	// CacheListening enables or disables KVStore listening at the cache layer
	CacheListening(listen bool)
}
```

```go
type CacheWrap interface {
	...

	// CacheWrapWithListeners recursively wraps again with listening enabled
	CacheWrapWithListeners(listeners []WriteListener) CacheWrap
}

type CacheWrapper interface {
	...

	// CacheWrapWithListeners recursively wraps again with listening enabled
	CacheWrapWithListeners(listeners []WriteListener) CacheWrap
}
```

> ### MultiStore implementation updates

Reviewer: ```suggestion

### MultiStore implementation updates
```

We will modify all of the `Store` and `MultiStore` implementations to satisfy these new interfaces, and adjust the `rootmulti` `GetKVStore` method
to wrap the returned `KVStore` with a `listenkv.Store` if listening is turned on.

```go
func (rs *Store) GetKVStore(key types.StoreKey) types.KVStore {
	store := rs.stores[key].(types.KVStore)

	if rs.TracingEnabled() {
		store = tracekv.NewStore(store, rs.traceWriter, rs.traceContext)
	}
	if rs.ListeningEnabled(key) {
		store = listenkv.NewStore(store, rs.listeners[key])
	}

	return store
}
```

We will also adjust the `cachemulti` constructor methods and the `rootmulti` `CacheMultiStore` method to enable listening

> in the cache layer when `CacheListening` is turned on.

Reviewer: Why do we need an explicit `CacheListening`? The root-multistore is always cache-wrapped during block execution. Shouldn't we always do this? Or am I missing something?

Author: No I think I'm missing something. The idea was that- assuming some Puts/Deletes in cache layer may never be written to a base Store- we would want to be able to turn on/off listening in the cache layer independently, so that we can choose to only listen to changes that are persisted, only changes in the cache layer, or both.

When the rootmulti is cache-wrapped [here](https://github.com/i-norden/cosmos-sdk/blob/multistore_state_listening/store/rootmulti/store.go#L451) the Stores it forwards are not yet listener-wrapped. Only if enabled will the cachemulti constructor [wrap the Stores in the cache layer with the listeners](https://github.com/i-norden/cosmos-sdk/blob/multistore_state_listening/store/cachemulti/store.go#L60).

Author: I think I see my mistake now (one of them, at least). The Stores should be listener wrapped before passing them to the cachemulti constructor. Otherwise there is no listening when cachemulti writes out the Put/Delete operations to the underlying Stores, instead it only listens to the Put/Deletes in the cache layer and only if`CacheListening` is turned on. I still wonder if it would be useful to enable independent listening in the cache and base layer?

```go
func (rs *Store) CacheMultiStore() types.CacheMultiStore {
	stores := make(map[types.StoreKey]types.CacheWrapper)
	for k, v := range rs.stores {
		stores[k] = v
	}
	var cacheListeners map[types.StoreKey][]types.WriteListener
	if rs.cacheListening {
		cacheListeners = rs.listeners
	}
	return cachemulti.NewStore(rs.db, stores, rs.keysByName, rs.traceWriter, rs.traceContext, cacheListeners)
}
```

> ### Exposing the data 

Reviewer: ```suggestion
### Exposing the data 

```

We will introduce and document mechanisms for exposing data from the above listeners to external consumers.

> #### Writing to file

Reviewer: ```suggestion
#### Writing to file

```

We will document and provide examples of how to configure a listener to write out to a file.
No new type implementation will be needed, an `os.File` can be used as the underlying `io.Writer` for a `GobWriteListener`.

Writing to a file is the simplest approach for streaming the data out to consumers.
This approach also provide the advantages of being persistent and durable, and the files can be read directly
or an auxiliary streaming services can tail the files and serve the data remotely.

Without pruning the file size can grow indefinitely, this will need to be managed by

> the developer in an application or even module-specific manner.

Reviewer: ```suggestion
the developer in an application or even module-specific manner (e.g. log rotation).
```

#### Writing to gRPC stream
We will implement and document an `io.Writer` type for exposing our listeners over a gRPC server stream.

Writing to a gRPC stream gRPC will allow us to expose the data over the standard gRPC interface.
This interface can be exposed directly to consumers or we can implement a message queue or secondary streaming service on top.

> Using gRPC will provide us with all of the regular advantages of gRPC and protobuf: versioning guarantees, client side code generation, and interoperability with the many gRPC plugins and auxillary services.

Reviewer: ```suggestion
Using gRPC will provide us with all of the regular advantages of gRPC and protobuf: versioning guarantees,
client side code generation, and interoperability with the many gRPC plugins and auxiliary services.
```

Proceeding through a gRPC intermediate will provide additional overhead, in most cases this is not expected to be rate limiting but in
instances where it is the developer can implement a more performant streaming mechanism for state listening.

### Configuration

> We will provide detailed documentation on how to configure the state listeners and their external streaming services from within an app's `AppCreator`,

Reviewer: ```suggestion

We will provide detailed documentation on how to configure the state listeners and their external streaming services from within an app's `AppCreator`,
```

using the provided `AppOptions`. We will add two methods to the `BaseApp` to enable this configuration:

```go
// SetCommitMultiStoreListeners sets the KVStore listeners for the provided StoreKey
func (app *BaseApp) SetCommitMultiStoreListeners(key sdk.StoreKey, listeners []storeTypes.WriteListener) {
	app.cms.SetListeners(key, listeners)
}

// SetCacheListening turns on or off listening at the cache layer
func (app *BaseApp) SetCacheListening(listening bool) {
	app.cms.CacheListening(listening)
}
```

As a demonstration, we will implement the state watching features as part of SimApp.
For example, the below is a very rudimentary integration of the state listening features into the SimApp `AppCreator` function:

> 

Reviewer: We need a brief section on what the actual (TOML) configuration options are.

e.g.

```toml
[store]
  listening = true|false
  # ...
```

Reviewer: We need to also probably support a listeners section:

```toml
[store]
  listeners = [
    "file",
    "grpc"
  ]
```

And we may not even need `listening` -- we can just assume enabled|disabled based off of `len(listeners) > 0`.

```go
func NewSimApp(
	logger log.Logger, db dbm.DB, traceStore io.Writer, loadLatest bool, skipUpgradeHeights map[int64]bool,
	homePath string, invCheckPeriod uint, encodingConfig simappparams.EncodingConfig,
	appOpts servertypes.AppOptions, baseAppOptions ...func(*baseapp.BaseApp),
) *SimApp {

	...

	keys := sdk.NewKVStoreKeys(
		authtypes.StoreKey, banktypes.StoreKey, stakingtypes.StoreKey,
		minttypes.StoreKey, distrtypes.StoreKey, slashingtypes.StoreKey,
		govtypes.StoreKey, paramstypes.StoreKey, ibchost.StoreKey, upgradetypes.StoreKey,
		evidencetypes.StoreKey, ibctransfertypes.StoreKey, capabilitytypes.StoreKey,
	)
	tkeys := sdk.NewTransientStoreKeys(paramstypes.TStoreKey)
	memKeys := sdk.NewMemoryStoreKeys(capabilitytypes.MemStoreKey)

	// configure state listening capabilities using AppOptions

> 	if cast.ToBool(appOpts.Get("simApp.listening")) {

Reviewer: ```suggestion
	if cast.ToBool(appOpts.Get("store.listening")) {
```

> 		writeDir := filepath.Clean(cast.ToString(appOpts.Get("simApp.listening.writeDir")))

Reviewer: ```suggestion
		writeDir := filepath.Clean(cast.ToString(appOpts.Get("store.listening.writeDir")))
```

real namespace TBD.

		for _, key := range keys {
			loadListener(bApp, writeDir, key)
		}
		for _, key := range tkeys {
			loadListener(bApp, writeDir, key)
		}
		for _, key := range memKeys {
			loadListener(bApp, writeDir, key)

> 		}

Reviewer: Why do we need to listen for writes on transient and mem keys? Transient keys aren't used for real data AFAICR and mem keys are only used for ephemeral data (specifically the x/capability store).

Author: Thank you, this was a misunderstanding on my part. In that case I will remove this.

> 		bApp.SetCacheListening(cast.ToBool(appOpts.Get("simApp.cacheListening")))

Reviewer: ```suggestion
		bApp.SetCacheListening(cast.ToBool(appOpts.Get("store.cacheListening")))
```

namespace TBD

	}
	
	...

	return app
}

> // loadListener creates and adds to the BaseApp a listener that writes out to a file in the provided write directory

Reviewer: I think we'll need to be able to map from `store.listeners` to helper functions. i.e. in app construction we look at all the listeners in the config and know which helper function to call`.

Author: Makes sense, I'll spec out a helper function lookup table. This would best fit as part of the `baseapp` package?

We will also introduce two approaches for exposing the data to external consumers: writing to files and writing to a gRPC stream.

### Listening interface
In a new file- `store/types/listening.go`- we will create a `Listening` interface for streaming out an allowed subset of state changes from a KVStore.
The interface can be backed by a simple wrapper around any underlying `io.Writer`.

```go
// Listening interface comprises the methods needed to filter and stream data out from a KVStore
type Listening interface {
	io.Writer // the underlying io.Writer
	Allowed(op Operation, key []byte) bool // method used to check if the Listener is allowed to listen to a specific state change
	TraceContext() TraceContext // method to access this Listener's TraceContext
}
```

We will lift some of the private types currently being used by the `tracekv.Store` into public types at a new location- `store/types/tracing.go`- for use in the `Listening` interface
```go
type Operation string

const (
	WriteOp     Operation = "write"
	ReadOp      Operation = "read"
	DeleteOp    Operation = "delete"
	IterKeyOp   Operation = "iterKey"

> 	IterValueOp Operation = "iterValue"

Reviewer 2: I see no use case for listening to anything besides writes and deletions. Maybe there's is one but it isn't explained and seems out or scope of this ADR

Author: I think there could be value to being able to watch the stateless operations, e.g. listening to `Get` operations to watch a node's read access. Happy to either expand on in the ADR that or adjust to focus on writes and deletes.

Reviewer 2: I mean maybe there's value in that but it's not actually watching for state changes so seems out of scope.

// it wraps around the underlying writer interface
func (l *Listener) Write(b []byte) (int, error) {
	return l.writer.Write(b)
}

// Allowed satisfies Listening interface
// it returns whether or not the Listener is allowed to listen to the provided operation at the provided key
func (l *Listener) Allowed(op Operation, key []byte) bool {
	// first check if the operation is allowed
	if _, ok := l.allowedOperations[op]; !ok {
		return false
	}
	// if there are no keys or prefixes in the whitelists then every key is allowed (unless disallowed in blacklists)
	// if there are whitelisted keys or prefixes then only the keys which conform are allowed (unless disallowed in blacklists)
	allowed := true
	if len(l.whitelistedKeys) > 0 || len(l.whitelistedPrefixes) > 0 {
		allowed = listsContain(l.whitelistedKeys, l.whitelistedPrefixes, key)
	}
	return allowed && !listsContain(l.blacklistedKeys, l.blacklistedPrefixes, key)
}

func listsContain(keys, prefixes [][]byte, key []byte) bool {
	for _, k := range keys {
		if bytes.Equal(key, k) {
			return true
		}
	}
	for _, p := range prefixes {
		if bytes.HasPrefix(key, p) {
			return true
		}
	}
	return false

> }

Reviewer 2: This should use maps rather than linear search.

Author: Sounds good, the trade off is that we need to convert to strings since we can't use slices as map keys.

Reviewer 2: So let me ask you this, for just the use case of caching to an external database will you be watching all keys or just some keys? i.e. is the use-case for watching some keys more related to gRPC subscribers?

Either way, I'm not really comfortable with watches on resources that are high churn being dealt with using linear search. Keep in mind that the way this is designed currently, it blocks writes on every single store write. This is a _very_ high cost.

I would say for the use case of caching to an external DB we really just want to capture _all_ writes with fault tolerance to some sort of on-disk WAL and do it fast to not block writing too much. If writing is blocked too much, the sync process simply won't be able to keep up when there's heavy usage. The sync process should then consume that WAL on a separate thread as quickly as possible.

For gRPC subscribers which I'm assuming might generally be web clients, I'm not sure fault tolerance is so important. Either way, the dispatch of all that data should also happen in separate threads so that writing is not blocked too much.

For any of these cases, if we're doing filtering for one client or many, I don't really think this is a use case where we settle for an inefficient design. In the web-client case, we should assume many clients, and in all cases we should imagine heavy writes. I actually think the web-client case should be dealt with separately because responding to many clients seems like its own problem space. In any event, I think we'll need to think beyond the restrictions of go `map`s and consider tries, etc. Maybe there are off the shelf solutions that would do that sort of subscription filtering for many clients. Anyway, for now maybe let's try to just get the database sync solution right?

Author: Thank you, that helps put things in perspective! Yeah for caching and indexing in a database we will be watching every key, I can think of hypothetical scenarios where we would only want to cache and index a subset of the data but that is both out of scope and, well, hypothetical.

For the stock  `WriteListener` I'll remove the filtering, If someone wants to filter they can implement a `WriteListener` that does so (using a more efficient approach for the prefixes than the one here) or they can perform the filtering at a higher level.

Reviewer 2: Great! I think that simplifies a lot what needs to be done for this first iteration, right?

Author: Definitely!

}

// Set implements the KVStore interface. It traces a write operation and
// delegates the Set call to the parent KVStore.
func (tkv *Store) Set(key []byte, value []byte) {
	types.AssertValidKey(key)
	writeOperation(tkv.listeners, types.WriteOp, key, value)
	tkv.parent.Set(key, value)
}

// Delete implements the KVStore interface. It traces a write operation and
// delegates the Delete call to the parent KVStore.
func (tkv *Store) Delete(key []byte) {
	writeOperation(tkv.listeners, types.DeleteOp, key, nil)
	tkv.parent.Delete(key)
}

// Has implements the KVStore interface. It delegates the Has call to the
// parent KVStore.
func (tkv *Store) Has(key []byte) bool {
	return tkv.parent.Has(key)
}

...

// writeOperation writes a KVStore operation to the underlying io.Writer of
// every listener that has permissions to listen to that operation at the given key
// The TraceOperation is JSON-encoded with the `key` and `value` fields as base64 encoded strings
func writeOperation(listeners []types.Listening, op types.Operation, key, value []byte) {
	// short circuit if there are no listeners so we don't waste time base64 encoding `key` and `value`
	if len(listeners) == 0 {
		return
	}

> 	traceOp := types.TraceOperation{

Reviewer 2: Why do things need to be base64 encoded? Can't we just pass are bytes to listeners?

Author: I was following the approach used in the `tracekv` store here, I'll adjust to pass raw bytes instead.

# ADR 038: KVStore state listening

## Changelog

- 11/23/2020: Initial draft

## Status

Proposed

## Abstract

This ADR defines a set of changes to enable listening to state changes of individual KVStores and exposing these data to consumers.

## Context

Currently, KVStore data can be remotely accessed through [Queries](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/messages-and-queries.md#queries) which proceed through Tendermint and the ABCI.
In addition to these request/response queries, it would be beneficial to have a means of listening to state changes as they occur in real time.

## Decision

We will modify the `MultiStore` interface and its concrete (`rootmulti` and `cachemulti`) implementations and introduce a new `listenkv.Store` to allow listening to state changes in underlying KVStores.

> We will also introduce two approaches for exposing the data to external consumers: writing to files and writing to a gRPC stream.

Reviewer 2: What is the use case for gRPC here? I'm not sure we need to include it for this approach.

# ADR 038: KVStore state listening

## Changelog

- 11/23/2020: Initial draft

## Status

Proposed

## Abstract

This ADR defines a set of changes to enable listening to state changes of individual KVStores and exposing these data to consumers.

## Context

Currently, KVStore data can be remotely accessed through [Queries](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/messages-and-queries.md#queries) which proceed through Tendermint and the ABCI.
In addition to these request/response queries, it would be beneficial to have a means of listening to state changes as they occur in real time.

## Decision

We will modify the `MultiStore` interface and its concrete (`rootmulti` and `cachemulti`) implementations and introduce a new `listenkv.Store` to allow listening to state changes in underlying KVStores.
We will also introduce the tooling for writing these state changes out to a file.

### Listening interface

> In a new file- `store/types/listening.go`- we will create a `WriteListener` interface for streaming out state changes from a KVStore.

Author: Missed this one!

```go
// WriteListener interface for streaming data out from a listenkv.Store
type WriteListener interface {
  // if value is nil then it was deleted

>   OnWrite(key []byte, value []byte)

Reviewer 3: would the block height be useful here?

Author: I think it would, but I don't think it is possible to emit that information from within the context of the KVStores without some significant refactoring.

Reviewer: Can we not pass in an `sdk.Context` as part of this? That way we can grab those fields (i.e the block header).

Author: I don't think this is possible without some significant refactoring, since this method will be called from within a KVStore which do not (currently) hold a reference to `sdk.Context`.

}
```

### Listener type

We will create two concrete implementation of the `WriteListener` interface in `store/types/listening.go`.

One that writes out length-prefixed key-value pairs to an underlying `io.Writer`:

```go
// PrefixWriteListener is used to configure listening to a KVStore by writing out big endian length-prefixed
// key-value pairs to an io.Writer
type PrefixWriteListener struct {
	writer io.Writer
	prefixBuf [6]byte
}

// NewPrefixWriteListener wraps a PrefixWriteListener around an io.Writer
func NewPrefixWriteListener(w io.Writer) *PrefixWriteListener {
	return &PrefixWriteListener{
		writer: w,
	}
}

// OnWrite satisfies the WriteListener interface by writing out big endian length-prefixed key-value pairs
// to an underlying io.Writer
// The first two bytes of the prefix encode the length of the key
// The last four bytes of the prefix encode the length of the value
// This WriteListener makes two assumptions
// 1) The key is no longer than 1<<16 - 1

> // 2) The value is no longer than 1<<32 - 1

Reviewer 3: why not just write:
```
[key-length-prefix][key-bytes][value-length-prefix][value-bytes]{repeat}
```

and we can remove the 2 assumptions here.

Author: I think we'd still need to define a static length for the key-length and value-length prefixes, and then assume that all keys and values fit within those lengths. Or we could use additional 1 byte prefix-length prefixes for the key and value prefixes to inform their length.

e.g.

`[key-length-prefix-length-prefix][key-length-prefix][key][value-length-prefix-length-prefix][value-length-prefix][value]`

Reviewer: I don't see why we'd need to do that. length-prefixed encoding should be fixed size, not variable. (e.g. int16)

Reviewer: You could also have this just be a protobuf-based message.

Author: To clarify further, it's not the key bytes or value bytes themselves that are overflowing here, its their length prefixes that can potentially overflow. If the key is longer than `1<<16 - 1` we can't encode its length in the two byte prefix, if the value is longer than `1<<32-1` we can't encode its length in the the four byte prefix.

Reviewer 3: During our regen call yesterday, varints were proposed to be used here as length prefixes. TBH this becomes similar to what protobuf does, so using protobuf also seems like a good idea to me (final binary size might be bigger, but we're already using protobuf everywhere so...)

Author: Thank you @alexanderbez and @amaurymartiny, protobuf it is :)

// OnWrite satisfies WriteListener interface by writing out newline delineated big endian key-length-prefixed key-value
// pairs to the underlying io.Writer
// The first two bytes encode the length of the key
// Separate key-value pairs are newline delineated
// This WriteListener makes three assumptions
// 1) The key is no longer than 1<<16 - 1
// 2) The value and keys contain no newline characters
func (l *NewlineWriteListener) OnWrite(key []byte, value []byte) {
	keyLen := len(key)
	if keyLen > math.MaxUint16 {
		return
	}
	binary.BigEndian.PutUint16(l.keyLenBuf[:], uint16(keyLen))
	l.writer.Write(e.keyLenBuf[:])
	l.writer.Write(key)
	l.writer.Write(value)
	l.writer.Write(newline)
}
```

The former makes no assumptions about the presence of newline characters in keys or values, but values
must be no longer than 1<<32 - 1. The latter assumes newlines are not present in keys or values but can support any length
of value. Both assume keys are no longer than 1<<16 - 1. Newline delineation improves readability by enabling a consumer to orient
themselves at the start of a key-value pair at any point in the stream (e.g. tail a file), without character delineation a consumer must start

> at the beginning of the stream and not lose track of their position in the stream.

Reviewer 3: As a counter argument, I would say this adds additional work. If there are `N` WriteListeners and `M` StreamingServices, that's potentially `N*M` connectors/adaptors to maintain.

What do you think of using this ADR to define one single spec for what the WriteListener writes, and consumers can get creative with StreamingServices by reading bytes from this spec.

Author: I like that idea, I feel like the scope creep is catching up on this ADR (my bad). 

As a compromise this ADR can be focused on the `WriteListener` and the single file streaming service that @aaronc outlines above. Using that approach to writing to files, we can abandon the need for newline delineation as instead delineation occurs at the file level, e.g. it's okay to have to always have to start at the head of the file if each file is only for a single ABCI stage of a single block.

		store = tracekv.NewStore(store, rs.traceWriter, rs.traceContext)
	}
	if rs.ListeningEnabled(key) {
		store = listenkv.NewStore(store, rs.listeners[key])
	}

	return store
}
```

We will also adjust the `cachemulti` constructor methods and the `rootmulti` `CacheMultiStore` method to forward the listeners
to and enable listening in the cache layer.

```go
func (rs *Store) CacheMultiStore() types.CacheMultiStore {
	stores := make(map[types.StoreKey]types.CacheWrapper)
	for k, v := range rs.stores {
		stores[k] = v
	}
	return cachemulti.NewStore(rs.db, stores, rs.keysByName, rs.traceWriter, rs.traceContext, rs.listeners)
}
```

### Exposing the data

We will introduce a new `StreamingService` interface for exposing `WriteListener` data streams to external consumers.

```go
// StreamingService interface for registering WriteListeners with the BaseApp and updating the service with the ABCI context
type StreamingService interface {
	Listeners() map[sdk.StoreKey][]storeTypes.WriteListener // returns the streaming service's listeners for the BaseApp to register
	BeginBlockReq(req abci.RequestBeginBlock) // update the streaming service with the latest RequestBeginBlock message

> 	BeginBlockResres abci.ResponseBeginBlock) // update the steaming service with the latest ResponseBeginBlock message

Reviewer 3: ```suggestion
	BeginBlockRes(res abci.ResponseBeginBlock) // update the steaming service with the latest ResponseBeginBlock message
```

	EndBlockReq(req abci.RequestEndBlock) // update the steaming service with the latest RequestEndBlock message

> 	EndBlockRes(res abci.ResponseEndBlock) // update the steaming service with the latest ResponseEndBlock message

Reviewer 3: Is this out of scope of this ADR, which as I understand only adds listeners on the store level?

Also, these listeners seems more appropriate on Tendermint.

Author: IMO it is in scope. This relates back to Aaron's latest comments, and to how we want to maintain the relation of state changes to the ABCI messages/block context that caused them. This is important for downstream IPLD-ization, we need to maintain the hash-linked nature of the data or be able to reconstitute it.

Reviewer 2: Yeah otherwise it's hard to know what block has been cached. They're just raw kv pairs with no time info. So even if we don't care about proofs we still need to commit the data to Postgres, etc transactionally.

	DeliverTxReq(req abci.RequestDeliverTx) // update the steaming service with the latest RequestDeliverTx message

>     DeliverTxRes(res abci.ResponseDeliverTx) // update the steaming service with the latest ResponseDeliverTx message

Reviewer 3: We're adding more computation to be run on each ABCI call. It would be good to have an idea of the performance overhead.

Author: Completely agree and I admittedly do not have a great feel for this yet, I could do some preliminary benchmarking. These methods would be updating the `StreamingService` with the latest ABCI message, so that it can e.g. generate a new file which retains that context and which we write out the state changes that are caused by that particular message.

}
```

#### Writing state changes to files

We will introduce an implementation of `StreamingService` which writes state changes out to a file.

```go
// FileStreamingService is a concrete implementation of StreamingService that writes state changes out to a file
type FileStreamingService struct {
	listeners map[sdk.StoreKey][]storeTypes.WriteListener // the listeners that will be initialized with BaseApp
	writeDir string
	filePrefix string
	fileSuffix string
	dst *os.File // the current write output file
}

```

Writing to a file is the simplest approach for streaming the data out to consumers.
This approach also provide the advantages of being persistent and durable, and the files can be read directly,
or an auxiliary streaming services can tail the files and serve the data remotely.

#### File pruning

Without pruning the file size can grow indefinitely, this will need to be managed by

> the developer in an application or even module-specific manner (e.g. log rotation).

Reviewer 4: very sensible approach indeed :+1:

# ADR 038: KVStore state listening

## Changelog

- 11/23/2020: Initial draft

## Status

Proposed

## Abstract

This ADR defines a set of changes to enable listening to state changes of individual KVStores and exposing these data to consumers.

## Context

Currently, KVStore data can be remotely accessed through [Queries](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/messages-and-queries.md#queries)
which proceed either through Tendermint and the ABCI, or through the gRPC server.

> In addition to these request/response queries, it would be beneficial to have a means of listening to state changes as they occur in real time.

Reviewer 5: Yes, and it would be great to enable many remote clients to listen to these changes...

In [another sdk](https://github.com/iov-one/weave), we wrote events on every set and delete. We didn't encode state, but it did provide a great way to subscribe to eg. changes in your bank balance

## Decision

We will modify the `MultiStore` interface and its concrete (`rootmulti` and `cachemulti`) implementations and introduce a new `listenkv.Store` to allow listening to state changes in underlying KVStores.
We will also introduce the tooling for writing these state changes out to files and configuring this service.

### Listening interface

In a new file, `store/types/listening.go`, we will create a `WriteListener` interface for streaming out state changes from a KVStore.

```go
// WriteListener interface for streaming data out from a listenkv.Store
type WriteListener interface {

> 	// if value is nil then it was deleted

Reviewer 5: It would be good to separate writing `[]byte{}` and deleting.
It is technically possible to write an empty value, and it does remain in the iavl store (eg. via iterators)

Reviewer 2: Hmm... I'm not sure this is true - I've tried that and gotten an error.

	// storeKey indicates the source KVStore, to facilitate using the the same WriteListener across separate KVStores
	OnWrite(storeKey types.StoreKey, key []byte, value []byte)
}
```

### Listener type

We will create a concrete implementation of the `WriteListener` interface in `store/types/listening.go`, that writes out protobuf
encoded KV pairs to an underlying `io.Writer`.

This will include defining a simple protobuf type for the KV pairs. In addition to the key and value fields this message
will include the StoreKey for the originating KVStore so that we can write out from separate KVStores to the same stream/file
and determine the source of each KV pair.

```protobuf
message StoreKVPair {
  optional string store_key = 1;
  required bytes key = 2;

>   required bytes value = 3;

Reviewer 5: You could add a type here (set/delete)

Author: Great idea, will do this to account for your above point!

// Operations are traced on each core KVStore call and written to any of the
// underlying listeners with the proper key and operation permissions
type Store struct {
	parent    types.KVStore
	listeners []types.WriteListener
	parentStoreKey types.StoreKey
}

// NewStore returns a reference to a new traceKVStore given a parent
// KVStore implementation and a buffered writer.
func NewStore(parent types.KVStore, psk types.StoreKey, listeners []types.WriteListener) *Store {
	return &Store{parent: parent, listeners: listeners, parentStoreKey: psk}
}

// Set implements the KVStore interface. It traces a write operation and
// delegates the Set call to the parent KVStore.
func (s *Store) Set(key []byte, value []byte) {
	types.AssertValidKey(key)
	s.parent.Set(key, value)
	s.onWrite(key, value)
}

// Delete implements the KVStore interface. It traces a write operation and
// delegates the Delete call to the parent KVStore.
func (s *Store) Delete(key []byte) {
	s.parent.Delete(key)
	s.onWrite(key, nil)
}

// onWrite writes a KVStore operation to all of the WriteListeners
func (s *Store) onWrite(key, value []byte) {
	for _, l := range s.listeners {

> 		l.OnWrite(s.parentStoreKey, key, value)

Reviewer 5: Doing this sync inside the Store means we must guarantee the listeners are fast.
Thus, disk writes may be too slow? I see the utility of not adding go-routines here, but listeners may want to just queue the data and let some other listener handle it in a goroutine.

Reviewer 5: (answered below)

Author: Going to think on this a bit more. We use a goroutine to queue the data but writing it out is still currently blocking in the BaseApp abci.go method.

Author: Ahh actually it's not really specified if it is blocking or not right now e.g. https://github.com/cosmos/cosmos-sdk/pull/8012/files#diff-96f43642682fd41beb81b2dad576c2149ae2a020389833462e7509775c4f51cbR302, will fill in some details. We could have options for either a synchronous or asynchronous `FileStreamingService`.

Author: An issue I'm seeing with writing the data out asynchronously is that if that process cannot keep up with the head, then the amount of data we are holding in memory can grow indefinitely. If it can keep up with head, then I think it might as well be done synchronously.

	}
}
```

### MultiStore interface updates

We will update the `MultiStore` interface to allow us to wrap a set of listeners around a specific `KVStore`.
Additionally, we will update the `CacheWrap` and `CacheWrapper` interfaces to enable listening in the caching layer.

```go
type MultiStore interface {
	...

	// ListeningEnabled returns if listening is enabled for the KVStore belonging the provided StoreKey
	ListeningEnabled(key StoreKey) bool

	// SetListeners sets the WriteListeners for the KVStore belonging to the provided StoreKey
	// It appends the listeners to a current set, if one already exists
	SetListeners(key StoreKey, listeners []WriteListener)
}
```

```go
type CacheWrap interface {
	...

	// CacheWrapWithListeners recursively wraps again with listening enabled

> 	CacheWrapWithListeners(storeKey types.StoreKey, listeners []WriteListener) CacheWrap

Reviewer 5: listeners here listen to changes in the cache, even if they are never written to the underlying store?

Author: My thinking was that this would work similarly to how the tracekv is used in the cache layer. The listenkv store will be wrapped around the parent kvstore in the cache kvstore (a la https://github.com/cosmos/cosmos-sdk/blob/master/store/iavl/store.go#L159) and then the listening would only occur when the cache layer updates are [written out to the underlying parent kvstore](https://github.com/cosmos/cosmos-sdk/blob/master/store/cachekv/store.go#L120).

to wrap the returned `KVStore` with a `listenkv.Store` if listening is turned on for that `Store`.

```go
func (rs *Store) GetKVStore(key types.StoreKey) types.KVStore {
	store := rs.stores[key].(types.KVStore)

	if rs.TracingEnabled() {
		store = tracekv.NewStore(store, rs.traceWriter, rs.traceContext)
	}
	if rs.ListeningEnabled(key) {
		store = listenkv.NewStore(key, store, rs.listeners[key])
	}

	return store
}
```

We will also adjust the `cachemulti` constructor methods and the `rootmulti` `CacheMultiStore` method to forward the listeners
to and enable listening in the cache layer.

```go
func (rs *Store) CacheMultiStore() types.CacheMultiStore {
	stores := make(map[types.StoreKey]types.CacheWrapper)
	for k, v := range rs.stores {
		stores[k] = v
	}
	return cachemulti.NewStore(rs.db, stores, rs.keysByName, rs.traceWriter, rs.traceContext, rs.listeners)
}
```

### Exposing the data

We will introduce a new `StreamingService` interface for exposing `WriteListener` data streams to external consumers.

```go
// Hook interface used to hook into the ABCI message processing of the BaseApp

> type Hook interface {

Reviewer 5: Very nice.
How are these messages interlaced to with the `WriteListener` messages?
To easily see which writes occured in EndBlock, which in a given tx, etc.

Reviewer 5: (answered below)

	ListenDeliverTx(ctx sdk.Context, req abci.RequestDeliverTx, res abci.ResponseDeliverTx) // update the steaming service with the latest DeliverTx messages
}

// StreamingService interface for registering WriteListeners with the BaseApp and updating the service with the ABCI messages using the hooks
type StreamingService interface {
	Stream(wg *sync.WaitGroup, quitChan <-chan struct{}) // streaming service loop, awaits kv pairs and writes them to some destination stream or file
	Listeners() map[sdk.StoreKey][]storeTypes.WriteListener // returns the streaming service's listeners for the BaseApp to register
	Hook
}
```

#### Writing state changes to files

We will introduce an implementation of `StreamingService` which writes state changes out to files as length-prefixed protobuf encoded `StoreKVPair`s.
This service uses the same `StoreKVPairWriteListener` for every KVStore, writing all the KV pairs from every KVStore
out to the same files, relying on the `StoreKey` field in the `StoreKVPair` protobuf message to later distinguish the source for each pair.

> The file naming schema is as such:

Reviewer 5: And you answered my question here. Very nice!

Author: :) credit goes to @aaronc  for this

* After every `BeginBlock` request a new file is created with the name `block-{N}-begin`, where N is the block number. All
subsequent state changes are written out to this file until the first `DeliverTx` request is received. At the head of these files,
  the length-prefixed protobuf encoded `BeginBlock` request is written, and the response is written at the tail.
* After every `DeliverTx` request a new file is created with the name `block-{N}-tx-{M}` where N is the block number and M
is the tx number in the block (i.e. 0, 1, 2...). All subsequent state changes are written out to this file until the next
`DeliverTx` request is received or an `EndBlock` request is received. At the head of these files, the length-prefixed protobuf
  encoded `DeliverTx` request is written, and the response is written at the tail.
* After every `EndBlock` request a new file is created with the name `block-{N}-end`, where N is the block number. All
subsequent state changes are written out to this file until the next `BeginBlock` request is received. At the head of these files,

>   the length-prefixed protobuf encoded `EndBlock` request is written, and the response is written at the tail.

Reviewer 5: Will this not create way too many files for the filesystem at some point. eg. 10k blocks * 100 tx = 1 million files in a dir.

I guess there can be cleanup, but why not just one file per block?

Author: Yeah this will make a ton of files, although perhaps one advantage of splitting a block into separate files for each ABCI request/response stage is that we can prune the files with that level of granularity. Additionally the separate files help direct our decoding of the binary data contained. E.g. for a `block-{N}-begin` file we know the first length-prefixed protobuf encoded entry is the `BeginBlock` request, that all subsequent length-prefixed entries are the resulting `StoreKVPair` state changes, until the final length-prefixed entry which is the response. If we mix, for example, `BeginBlock` and `DeliverTx` stages in the same file I think we would then need additional structure to tell us e.g. when the `BeginBlock` state changes are finished and the `BeginBlock` response begins since the `BeginBlock` response is no longer at the tail of the file. I'm nt sure that any of this is good enough justification for the additional file clutter, what do you think?

Reviewer 5: I think one file per block is plenty for granularity. Unless you need to start processing state changes from some txs before the endblock is written, I see no downside for grouping them 1/block.

It would need a clear file-format, as you mentioned. Ideally something that could be streamed

Reviewer 2: Increased granularity allows the sync process to work faster IMHO. The idea is really that these logs just get deleted as soon as they are consumed. It's basically a very dumb durable message queue. A single file per block could also work, we'd just need to specify the format better.

```go
// FileStreamingService is a concrete implementation of StreamingService that writes state changes out to a file
type FileStreamingService struct {
	listeners map[sdk.StoreKey][]storeTypes.WriteListener // the listeners that will be initialized with BaseApp

> 	srcChan <-chan []byte // the channel that all of the WriteListeners write their out to

Reviewer 5: You're a step ahead - nice with the chan implementation

	// create a new file with the req info according to naming schema
	// NOTE: if the tx failed, handle accordingly
	// write req to file
	// write all state changes cached for this stage to file
	// reset cache
	// write res to file
	// close file
}

// Stream spins up a goroutine select loop which awaits length-prefixed binary encoded KV pairs and caches them in the order they were received
func (fss *FileStreamingService) Stream(wg *sync.WaitGroup, quitChan <-chan struct{}) {
	wg.Add(1)
	go func() {
		defer wg.Done()
		for {
			select {
			case <-quitChan:
				return
                        case by := <-fss.srcChan:
                        	append(fss.stateCache, by)
			}
		}
	}()
}
```

Writing to a file is the simplest approach for streaming the data out to consumers.
This approach also provide the advantages of being persistent and durable, and the files can be read directly,
or an auxiliary streaming services can read from the files and serve the data over a remote interface.

#### Auxiliary streaming service

We will create a separate standalone process that reads and internally queues the state as it is written out to these files

> and serves the data over a gRPC API. This API will allow filtering of requested data, e.g. by block number, block/tx hash, ABCI message type,

Reviewer 5: grpc allows streaming responses.
It would be nice to allow subscriptions via such.

Not just searching for when changes happened to key X, but getting an event each time that happens (with the context: DeliverTx/Begin/End block)

Author: Completely agree, adding language to include subscriptions!

type StreamingService interface {
	Stream(wg *sync.WaitGroup, quitChan <-chan struct{}) // streaming service loop, awaits kv pairs and writes them to some destination stream or file
	Listeners() map[sdk.StoreKey][]storeTypes.WriteListener // returns the streaming service's listeners for the BaseApp to register
	Hook
}
```

#### Writing state changes to files

We will introduce an implementation of `StreamingService` which writes state changes out to files as length-prefixed protobuf encoded `StoreKVPair`s.
This service uses the same `StoreKVPairWriteListener` for every KVStore, writing all the KV pairs from every KVStore
out to the same files, relying on the `StoreKey` field in the `StoreKVPair` protobuf message to later distinguish the source for each pair.

The file naming schema is as such:
* After every `BeginBlock` request a new file is created with the name `block-{N}-begin`, where N is the block number. All
subsequent state changes are written out to this file until the first `DeliverTx` request is received. At the head of these files,

>   the length-prefixed protobuf encoded `BeginBlock` request is written, and the response is written at the tail.

Reviewer 2: How do you separate the response from the state changes when parsing?

Author: In this current format the only way to distinguish the response is that it is the final length-prefix encoded protobuf entry in the file. This can work if we buffer the bytes of each file fully into memory, segmenting it into separate protobuf objects by the length-prefixes, whereby we can recognize which object is the last before we attempt to decode it.

This does feel kinda janky! But I think it could be feasible since each file is just one ABCI stage of a Tendermint block, we already need all these bytes to fit in memory since we have to hold it all in memory before we flush it out to the file in the first place.  Do you think it'd be better to add something like a total-length prefix to the beginning of the state changes? E.g. a variable length prefix in front of the first length-prefixed state change that encodes the entire length of all the prefixed state changes.

	// create a new file with the req info according to naming schema
	// write req to file
	// write all state changes cached for this stage to file
	// reset cache
	// write res to file
	// close file
}

func (fss *FileStreamingService) ListenDeliverTx(ctx sdk.Context, req abci.RequestDeliverTx, res abci.ResponseDeliverTx) {
	// NOTE: this could either be done synchronously or asynchronously
	// create a new file with the req info according to naming schema
	// NOTE: if the tx failed, handle accordingly
	// write req to file
	// write all state changes cached for this stage to file
	// reset cache
	// write res to file
	// close file
}

// Stream spins up a goroutine select loop which awaits length-prefixed binary encoded KV pairs and caches them in the order they were received
func (fss *FileStreamingService) Stream(wg *sync.WaitGroup, quitChan <-chan struct{}) {
	wg.Add(1)
	go func() {
		defer wg.Done()
		for {
			select {
			case <-quitChan:
				return
                        case by := <-fss.srcChan:
                        	append(fss.stateCache, by)
			}
		}
	}()
}
```

Writing to a file is the simplest approach for streaming the data out to consumers.

> This approach also provide the advantages of being persistent and durable, and the files can be read directly,

Reviewer 2: ```suggestion
This approach also provides the advantages of being persistent and durable, and the files can be read directly,
```

or an auxiliary streaming services can read from the files and serve the data over a remote interface.

#### Auxiliary streaming service

We will create a separate standalone process that reads and internally queues the state as it is written out to these files

> and serves the data over a gRPC API. This API will allow filtering of requested data, e.g. by block number, block/tx hash, ABCI message type,

Reviewer 2: Is this for clients or caching to a database? Feels like the cache process should just read the files and talk to the DB and skip grpc. Or we should use an actual message queue instead of files + grpc.

Author: I think in the case of downstream IPLD-ization in Postgres we would just consume the files directly as we know we want to consume everything- as quickly as possible- and will run that process in the same environment with direct access to the files. But if we want to be more selective about what we consume and consume it remotely than that's where this auxiliary streaming service comes into play, as I see it. @robert-zaremba may be able to back me up here.

Author: To actually answer your question: this would be useful primarily for clients but also potentially database caching in those cases where the caching wants to leverage this more selective interface and/or perform the caching remotely.

Reviewer 6: Yes, it's for clients - eg data warehousing / ETL. We were discussing queues but didn't make a decision. Such a service will only need to expose a network port, without exposing filesystem.

	}
	if rs.ListeningEnabled(key) {
		store = listenkv.NewStore(key, store, rs.listeners[key])
	}

	return store
}
```

We will also adjust the `cachemulti` constructor methods and the `rootmulti` `CacheMultiStore` method to forward the listeners
to and enable listening in the cache layer.

```go
func (rs *Store) CacheMultiStore() types.CacheMultiStore {
	stores := make(map[types.StoreKey]types.CacheWrapper)
	for k, v := range rs.stores {
		stores[k] = v
	}
	return cachemulti.NewStore(rs.db, stores, rs.keysByName, rs.traceWriter, rs.traceContext, rs.listeners)
}
```

### Exposing the data

We will introduce a new `StreamingService` interface for exposing `WriteListener` data streams to external consumers.

```go
// StreamingService interface for registering WriteListeners with the BaseApp and updating the service with the ABCI context
type StreamingService interface {
	Stream(wg *sync.WaitGroup, quitChan <-chan struct{}) // streaming service loop, awaits kv pairs and writes them to some destination stream or file
	Listeners() map[sdk.StoreKey][]storeTypes.WriteListener // returns the streaming service's listeners for the BaseApp to register

> 	BeginBlockReq(req abci.RequestBeginBlock) // update the streaming service with the latest RequestBeginBlock message

Reviewer 2: I would also include the response types, and call all of these ListenBeginBlock, etc...
